File: 백준/Gold/13023. ABCDE/ABCDE.cc
#include <iostream>
#include <vector>

using namespace std;

int N, M;
int visited[2001];
vector<int> graph[2001];
bool flag = false;

void FastIO() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
}

void Init() {
    int a, b;
    cin >> N >> M;
    for (int i = 0; i < M; i++) {
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }
}

void dfs(int cur, int depth) {
    if (depth == 4) {
        flag = true;
        return;
    }
    visited[cur] = 1;
    for (int i = 0; i < graph[cur].size(); i++) {
        int next = graph[cur][i];
        if (!visited[next]) {
            dfs(next, depth + 1);
        }
    }
    visited[cur] = 0;
}

void solve() {
    for (int i = 0; i < M; i++) {
        fill(&visited[0], &visited[0] + 2001, 0);
        dfs(i, 0);
        if (flag) {
            cout << "1";
            return;
        }
    }
    cout << "0";
}

int main(void) {
    FastIO();
    Init();
    solve();
    return 0;
}

File: 백준/Gold/13023. ABCDE/ABCDE.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static int N, M;
    static boolean flag = false;
    static StringTokenizer st;
    static ArrayList<Integer>[] graph = new ArrayList[2001];
    static Set<Integer> vertex = new HashSet<>();
    static boolean[] visited;
    public static void main(String[] args) throws IOException {
        Init();
        solve();
    }

    public static void Init() throws IOException {
        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        for(int i = 0; i < 2001; i++){
            graph[i] = new ArrayList<>();
        }

        for(int i = 0; i < M; i++){
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            vertex.add(a);
            vertex.add(b);
            graph[a].add(b);
            graph[b].add(a);
        }
    }

    public static void dfs(int v, int cnt){
        visited[v] = true;
        if(flag)
            return;
        if(cnt == 4){
            flag = true;
            return;
        }

        for(int i = 0; i < graph[v].size(); i++){
            int next = graph[v].get(i);
            if(!visited[next]){
                dfs(next, cnt + 1);
            }
        }
        visited[v] = false;
    }

    public static void solve(){
        for(Integer v : vertex){
            if(flag)
                break;
            visited = new boolean[2001];
            dfs(v, 0);
        }
        if (flag)
            System.out.print("1");
        else {
            System.out.print("0");
        }
    }

}